### ISSUE TITLE
distributed.joblib do not runs while joblib does or How to execute an arbitrary code on the worker right before sending him a taks for deserialization?

### ISSUE THREAD

**Issue by thoth291** [dask/distributed#1082]
I've found that ipyparallel's
```python
dview.execute('import sys; sys.path.append("/shared/dir/with/source/code")')
```
and similar hacks using execute are quite helpful to setup environment before sending a task.
This way things like deserialization won't be messed up and will, probably, do the right thing.

Currently I'm dealing with a simple joblib program:
```python
nloops=5
wavelet = makeRicker(nt=1024,dt=0.004,f=30)
pinv = [Perturb(nt=1024,dt=0.004,nOffsets=256,dx=25,motionType=2,srcDepth=10,rvrDepth=10) for m in range(nloops)]
misfit=np.zeros(nm)

with parallel_backend('dask.distributed', scheduler_host='scheduler:8786'):
    pmisfit = Parallel(n_jobs=-1)(map(delayed(superf), range(nloops)))
```
So when I run it I get this:
```python
/dist/anaconda/lib/python2.7/site-packages/distributed/protocol/pickle.pyc in loads()
     57 def loads(x):
     58     try:
---> 59         return pickle.loads(x)
     60     except Exception:
     61         logger.info("Failed to deserialize %s", x[:10000], exc_info=True)

ImportError: No module named spie1d
```
Missing module is in that `/shared/dir/with/source/code` and appears in `superf` cause it using `pinv` and `wavelet` global variables.

The very same example with `multiprocessing` backend works fine with `joblib` so I'm a little bit puzzled why `distributed.joblib` is not working and that's why I think that having a `pre_task_accept closure in the workers class` or `cluster.execute` method would be useful.

Could you please advise what should I do in cases like that?

P.S. the code in that `/shared/dir/with/source/code` is not pure python - it has precompiled python modules (.so files - interfaces to some Fortran libs).

**Comment by thoth291** [comment]
I just realized that there is `--preload` option in `dask-worker` - so I guess I wonder how to do this per Parallel task.

**Comment by jcrist** [comment]
The distributed joblib backend just runs on a dask-distributed cluster. Any of the normal ways of making python libraries available to the workers should work fine here. A few possibilities:

- Actually install your python code into PYTHONPATH before starting your worker processes (recommended). Is there a reason you can't do this?
- Use [`Client.run`](http://distributed.readthedocs.io/en/latest/api.html#distributed.client.Client.run) and do what you're doing above:

```python
from dask.distributed import Client
c = Client(...)

def f():
    import sys;
    sys.path.append("/shared/dir/with/source/code")

c.run(f)
```
- Use [`Client.upload_file`](http://distributed.readthedocs.io/en/latest/api.html#distributed.client.Client.upload_file) to upload local python files to the worker, and place them on the worker's path. See [this stackoverflow question](http://stackoverflow.com/questions/39295200/can-i-use-functions-imported-from-py-files-in-dask-distributed) for more info.

**Comment by thoth291** [comment]
Thank you for your answer, @jcrist!
I'm glad I asked this question - now I see that dask.distributed is as powerful as ipython (in the sense how I use it).
I'll test on my problem soon and will report on the progress.

In the meanwhile, this only answers to the half of my question.
<hr>
The second half is that `JobLib` can handle this parallelization with `multiprocessing` backend while `distributed.joblib` can't. Is that by design and always will be like that for cases like this - or it's an accident/bug?

**Comment by jcrist** [comment]
> The second half is that `JobLib` can handle this parallelization with `multiprocessing` backend while `distributed.joblib` can't.

Joblib recently refactored how they do backends to allow for other (non-joblib) libraries to create their own. The `dask.distributed` backend is one, while the `multiprocessing`, `threading`, and `sequential` backends (which are included in `joblib`) are others. If you want `multiprocessing`, use the `multiprocessing` backend, if you want to parallelize across a cluster, use the `dask.distributed` backend. Does that answer your question?

**Comment by thoth291** [comment]
> Does that answer your question?
I'm afraid it's not.
Does it means that JobLib Parallel is not backend agnostic?
So if something works with `multiprocessing` - then I shouldn't expect it to work with `dask.distributed`?

P.S. Thank you so much for explaining me all of these!

**Comment by jakirkham** [comment]
Things have changed significantly over the past few years. If you have time and interest, would encourage you to upgrade and retry @thoth291 🙂

**Comment by GenevieveBuckley** [comment]
Hi @thoth291, I don't know if you have had a chance to try John's suggestion. If you do that and have more questions, I encourage you to open a new issue to discuss :smile:
